### ISSUE TITLE
Error while running helloworld example

### ISSUE THREAD

**Issue by namrata-ibm** [grpc/grpc#13458]
### What version of gRPC and what language are you using? 
- grpc v1.7.2
  
### What operating system (Linux, Windows, …) and version?
 - Linux s390x Ubuntu 16.04 
  
### What runtime / compiler are you using (e.g. python version or version of gcc) 
- gcc version 5.4.0 
  
### What did you do?
Built grpc 1.7.2 from source and ran `helloworld` example
 
### What did you expect to see?
 hello world output
 
### What did you see instead?
```
test@c379f028000f:~/grpc/examples/cpp/helloworld$ ./greeter_server
E1120 06:05:52.276953417    4630 sockaddr_utils.c:239]       Unknown socket family 0 in grpc_sockaddr_get_port
E1120 06:05:52.277093292    4630 tcp_server_utils_posix_common.c:90] assertion failed: port > 0
Aborted (core dumped)
``` 
### Anything else we should know about your project / environment?

grpc 1.7.2 on s390x along with protobuf 3.4.0 built successfully, however while running the `hello world` example, i am facing the error mentioned above. I also tried running unit_tests in c, observed multiple failures with the same error.

Part of test log:
```
bins/dbg/server_chttp2_test(+0x2d404)[0x2aa007ed404]
FAILED: bins/dbg/server_chttp2_test GRPC_POLL_STRATEGY=epollsig [ret=-6, pid=53558]

D1120 09:30:25.593451605   53561 test_config.c:388]          test slowdown factor: sanitizer=1, fixture=1, poller=5, total=5
D1120 09:30:25.593996668   53561 ev_posix.c:107]             Using polling engine: poll-cv
D1120 09:30:25.594065543   53561 dns_resolver.c:301]         Using native dns resolver
I1120 09:30:25.594153293   53561 cancel_with_status.c:42]    Running test: test_invoke_simple_request/chttp2/fullstack+load_reporting [1 ops]
E1120 09:30:25.674839604   53561 sockaddr_utils.c:239]       Unknown socket family 0 in grpc_sockaddr_get_port
E1120 09:30:25.674860291   53561 tcp_server_utils_posix_common.c:90] assertion failed: port > 0
```

Any pointers in resolving this would be helpful.

**Comment by namrata-ibm** [comment]
Further analysis shows that an older commit `d7ff4ff` was working fine on s390x. The server started successfully:
```
test@f8c8424fc10f:~ /grpc/examples/cpp/helloworld$ ./greeter_server &
[1] 27329
 Server listening on 0.0.0.0:50051
```

After tracking further, observed that gRPC is working correctly till `objective-c-v1.0.2` tag, however failing in `v1.1.0-pre1`. Checked various intermediate commits like  'https://github.com/grpc/grpc/commit/8cbaeb89f399f4e87a91aca6ac5d59b5a753dfb0' where the failure is seen. However not able to reach the cause of this behavior.

Anyone has any suggestions? @vjpai, could you please have a look?

**Comment by namrata-ibm** [comment]
@mehrdada , @jtattermusch .. Could you help with this?

**Comment by mehrdada** [comment]
@namrata-ibm I wonder if the address family AF_INET and AF_INET6 are defined correctly on your platform. It's hard for me to poke around since I don't have access to your architecture. :(

**Comment by namrata-ibm** [comment]
@mehrdada, Found the root cause of this failure. The issue is happening [here](https://github.com/grpc/grpc/blob/781fd6f6ea03645a520cd5c675da67ab61f87e4b/src/core/lib/iomgr/tcp_server_utils_posix_common.c#L185) while casting `sockname_temp.len` value.
I could see #11390 already addressing this issue. Could you please have a look at the PR and help in resolving this?

**Comment by mehrdada** [comment]
So you think the issue is https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/resolve_address.h#L30 should have been `socklen_t` instead of `size_t`? The PR has gone stale so we can't just merge that one.

@yashykt, I'm handing this over to you: can you please help to see what the implications of such change would be?

**Comment by namrata-ibm** [comment]
@mehrdada , yup, changing to `socklen_t` works on s390x(big endian).
@yashykt , could you please help with this?

**Comment by yashykt** [comment]
socklen_t might not be supported on all systems. Creating PR #14464 to verify if it works on all supported platforms.

**Comment by markdroth** [comment]
How exactly do the types of `size_t` and `socklen_t` differ on big-endian platforms?

If the problem here is endianness, don't we need to use `htonl()` and `ntohl()` macros instead of just casting?

**Comment by yashykt** [comment]
The issue over here is to do with different sizes of size_t and socklen_t
on some big endian machines. If we accept the address length in a size_t by
pointing to it with a socklen_t then we are basically filling the wrong
bytes.

**Comment by vjpai** [comment]
I don't think it sounds like endianness, but rather just operating system
structure sizes. Which is fine and important.

**Comment by yashykt** [comment]
As discussed in PR #14464, it is an endianness issue.

@namrata-ibm Can you verify that #14464 solves the issue?

**Comment by namrata-ibm** [comment]
Changes suggested in #14464 resolve the issues on s390x. 
Tried running `helloworld` example and also few tests from c module which pass.(tried in gRPC v1.8.3)
Thanks @yashykt  for your help.

**Comment by yashykt** [comment]
Thanks for verifying @namrata-ibm!

**Comment by yashykt** [comment]
Fixed in #14464
